### PR TITLE
feat(snap): add go-based configure and install hooks

### PR DESCRIPTION
**Pull request by tonyespy**
This commit adds go-based configure and install hooks to the
snap which adds additional support for environment variable
configuration overrides (see snap/README.md for details).

This commit also adds a content interface plug to allow
configuration.toml, devices, and profiles to be provided
by an external snap.

Signed-off-by: Tony Espy <espy@canonical.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [NA] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/device-camera-go/blob/master/.github/CONTRIBUTING.md

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## New Imports

- [ ] Yes
- [x] No

### CODE REVIEW

File: snap/local/hooks/cmd/configure/configure.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+ */
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	local "github.com/canonical/edgex-device-camera-go/hooks"
```
Comment (siggiskulason) [inline snap/local/hooks/cmd/configure/configure.go:26]: should be "github.com/edgexfoundry/device-camera-go/hooks"

File: snap/local/hooks/cmd/configure/configure.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	cli := hooks.NewSnapCtl()
+	envJSON, err = cli.Config(hooks.EnvConfig)
+	if err != nil {
+		hooks.Error(fmt.Sprintf("Reading config 'env' failed: %v", err))
+		os.Exit(1)
+	}
+
+	if envJSON != "" {
+		hooks.Debug(fmt.Sprintf("edgex-device-camera:configure: envJSON: %s", envJSON))
+		err = hooks.HandleEdgeXConfig("device-camera-go", envJSON, local.ConfToEnv)
```
Comment (siggiskulason) [inline snap/local/hooks/cmd/configure/configure.go:61]: should be `device-camera` (not `device-camera-go`)

File: snap/local/hooks/cmd/configure/configure.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	autostart = strings.ToLower(autostart)
+
+	hooks.Debug(fmt.Sprintf("edgex-device-camera autostart is %s", autostart))
+
+	// service is stopped/disabled by default in the install hook
+	switch autostart {
+	case "true":
+		fallthrough
+	case "yes":
+		err = cli.Start("device-camera-go", true)
```
Comment (siggiskulason) [inline snap/local/hooks/cmd/configure/configure.go:89]: should be `device-camera` (not `device-camera-go`)

File: snap/local/hooks/cmd/install/install.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+)
+
+var cli *hooks.CtlCli = hooks.NewSnapCtl()
+
+// installProfiles copies the profile configuration.toml files from $SNAP to $SNAP_DATA.
+func installConfig() error {
+	var err error
+
+	path := "/config/device-camera-go/res/configuration.toml"
+	destFile := hooks.SnapData + path
```
Comment (siggiskulason) [inline snap/local/hooks/cmd/install/install.go:36]: should be `/config/device-camera/configuration.toml` (not `device-camera-go`)

File: snap/local/hooks/cmd/install/install.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+// TODO: merge into the above function...
+func installDevProfiles() error {
+	var err error
+
+	profs := [...]string{"-axis", "-bosch", ""}
+
+	for _, v := range profs {
+		path := fmt.Sprintf("/config/device-camera-go/res/camera%s.yaml", v)
+		destFile := hooks.SnapData + path
```
Comment (siggiskulason) [inline snap/local/hooks/cmd/install/install.go:58]: `config/device-camera/`

File: snap/local/hooks/cmd/install/install.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	err = installDevProfiles()
+	if err != nil {
+		hooks.Error(fmt.Sprintf("edgex-device-camera:install: %v", err))
+		os.Exit(1)
+	}
+
+	// disable the service and handle the autostart logic in the configure hook
+	// as default snap configuration is not available when the install hook runs
+	// TODO: re-name service and drop "-go" (will happen for ireland)
+	err = cli.Stop("device-camera-go", true)
```
Comment (siggiskulason) [inline snap/local/hooks/cmd/install/install.go:97]: `device-camera`

File: snap/README.md [hunk 7]
```diff
@@ -32,3 +68,92 @@ ensures that as well as starting the service now, it will be automatically start
...
+The following syntax is used to specify service-specific configuration overrides:
+
+
+```env.<stanza>.<config option>```
+For instance, to setup an override of the service's Port use:
+```$ sudo snap set edgex-device-camera env.service.port=2112```
+And restart the service:
+```$ sudo snap restart edgex-device-camera.device-camera```
+**Note** - at this time changes to configuration values in the [Writable] section are not supported.
+For details on the mapping of configuration options to Config options, please refer to "Service Environment Configuration Overrides".
```
Comment (siggiskulason) [inline snap/README.md:132]: In Hanoi we had the following to set camera username/password
```
[Driver]
driver.user
driver.password
driver.auth-method
```

but that has now been moved into the `Writable` section. Is it worth mentioning something about that?

**Review by siggiskulason — CHANGES_REQUESTED** [review]
Looks good, but a few minor inline comments/change requests

File: snap/local/hooks/go.mod [hunk 8]
```diff
@@ -0,0 +1,5 @@
+module github.com/edgexfoundry/edgex-device-camera-go/hooks
```
Comment (siggiskulason) [inline snap/local/hooks/go.mod:1]: Should be `github.com/edgexfoundry/device-camera-go/hooks`

**Comment by siggiskulason** [comment]
I think that you might also need to update `snap/local/runtime-helpers/bin/startup-env-var.sh `

File: snap/README.md [hunk 9]
```diff
@@ -5,6 +5,42 @@ This folder contains snap packaging for the EdgeX Camera Device Service Snap
...
+$ snap info edgex-device-camera
+```
+
+The latest stable version of the snap can be installed using:
+
+```bash
+$ sudo snap install edgex-device-camera
+```
+
+The 2.0 (Ireland) release of the snap can be instaing using:
```
Comment (siggiskulason) [inline snap/README.md:31]: `instaing` => `installed`

**Comment by tonyespy** [comment]
Addressed all comments and force pushed...

**Review by siggiskulason — APPROVED** [review]
LGTM

**Pull request merged**